diff --git a/hints/darwin.sh b/hints/darwin.sh
index 6127aa1..9267791 100644
--- a/hints/darwin.sh
+++ b/hints/darwin.sh
@@ -368,6 +368,15 @@ ccflags="$ccflags -DNO_THREAD_SAFE_QUERYLOCALE"
 # #21556]
 ccflags="$ccflags -DNO_POSIX_2008_LOCALE"
 
+# See comments in locale.c about this #define
+ccflags="$ccflags -DHAS_BROKEN_LANGINFO_CODESET"
+
+# Get: "Assertion failed: (p->val == key), function lookup_substsearch, file
+# collate.c, line 596."
+if test "$darwin_major" -ge 24 && test "$perl_revision" -ge 5 && ( test "$perl_version" -ge 42 || ( test "$perl_version" -eq 41 && test "$perl_subversion" -ge 11 ) ); then
+    ccflags="$ccflags -DNO_LOCALE_COLLATE"
+fi
+
 ldlibpthname='DYLD_LIBRARY_PATH';
 
 # useshrplib=true results in much slower startup times.
@@ -424,11 +433,13 @@ EOM
     case `uname -p` in 
     powerpc) arch=ppc64 ;;
     i386) arch=x86_64 ;;
+    arm) arch=arm64 ;;
     *) cat <<EOM >&4
 
 *** Don't recognize processor, can't specify 64 bit compilation.
 
 EOM
+     exit 1
     ;;
     esac
     for var in ccflags cppflags ld ldflags
