diff --git a/hints/darwin.sh b/hints/darwin.sh
index 72bdfb6..9267791 100644
--- a/hints/darwin.sh
+++ b/hints/darwin.sh
@@ -287,14 +287,14 @@ case "$osvers" in  # Note: osvers is the kernel version, not the 10.x
    ldflags="${ldflags} -flat_namespace"
    lddlflags="${ldflags} -bundle -undefined suppress"
    ;;
-[7-9].*)   # OS X 10.3.x - 10.5.x
+[7-8].*)   # OS X 10.3.x - 10.4.x
    lddlflags="${ldflags} -bundle -undefined dynamic_lookup"
    case "$ld" in
        *MACOSX_DEPLOYMENT_TARGET*) ;;
        *) ld="env MACOSX_DEPLOYMENT_TARGET=10.3 ${ld}" ;;
    esac
    ;;
-*)        # OS X 10.6.x - current
+*)        # OS X 10.5.x - current
    # The MACOSX_DEPLOYMENT_TARGET is not needed,
    # but the -mmacosx-version-min option is always used.
 
@@ -357,13 +357,26 @@ EOM
         esac
     fi
 
-    # The OS is buggy with respect to this.
-    ccflags="$ccflags -DNO_POSIX_2008_LOCALE"
-
    lddlflags="${ldflags} -bundle -undefined dynamic_lookup"
    ;;
 esac
 
+# Darwin's querylocale() has races
+ccflags="$ccflags -DNO_THREAD_SAFE_QUERYLOCALE"
+
+# But it doesn't much matter because the whole implementation has bugs [GH
+# #21556]
+ccflags="$ccflags -DNO_POSIX_2008_LOCALE"
+
+# See comments in locale.c about this #define
+ccflags="$ccflags -DHAS_BROKEN_LANGINFO_CODESET"
+
+# Get: "Assertion failed: (p->val == key), function lookup_substsearch, file
+# collate.c, line 596."
+if test "$darwin_major" -ge 24 && test "$perl_revision" -ge 5 && ( test "$perl_version" -ge 42 || ( test "$perl_version" -eq 41 && test "$perl_subversion" -ge 11 ) ); then
+    ccflags="$ccflags -DNO_LOCALE_COLLATE"
+fi
+
 ldlibpthname='DYLD_LIBRARY_PATH';
 
 # useshrplib=true results in much slower startup times.
@@ -420,11 +433,13 @@ EOM
     case `uname -p` in 
     powerpc) arch=ppc64 ;;
     i386) arch=x86_64 ;;
+    arm) arch=arm64 ;;
     *) cat <<EOM >&4
 
 *** Don't recognize processor, can't specify 64 bit compilation.
 
 EOM
+     exit 1
     ;;
     esac
     for var in ccflags cppflags ld ldflags
